### ISSUE TITLE
JSONObject.toString does not use an Enum's `toString()` but its `name()`

### ISSUE THREAD

**Issue by arthurBricq** [stleary/JSON-java#838]
When calling the `toString` method of a `JSONObject`, if a field of the object is an enum, the `name` method is called (when constructing the string) instead of the `toString` method. This is problematic the  `name()` method is not overidable (as it is decalred final). 

I have found [where this is in the sources](https://github.com/stleary/JSON-java/blame/6dba7220e1c385e88ea00b70e11c2f0e65d4fe95/src/main/java/org/json/JSONObject.java#L2656).

As specified by the [java doc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Enum.html#name()) of the `name` function : 

> Most programmers should use the [toString()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Enum.html#toString()) method in preference to this one, as the toString method may return a more user-friendly name

The fix should be really easy, I can do it if it helps you. Is this an issue not tracked ? I have not found corresponding issues.

Here's a quick way to reproduce this default.

```java
public class TestEnumToString {
    enum MyEnum {
        V_1, V_2;

        @Override
        public String toString() {
            switch (this) {
                case V_1:
                    return "1.0";
                case V_2:
                    return "2.0";
            };
            return "";
        }
    }

    public static void main(String[] args) throws JSONException {
        JSONObject json = new JSONObject();
        json.put("v1", MyEnum.V_1);
        json.put("v2", MyEnum.V_2);
        System.out.println(json.toString());
    }
}
```

Thanks in advance, 
Arthur

**Comment by ThestralWarrior** [comment]
Hi can I work on this?

**Comment by stleary** [comment]
@ThestralWarrior Sure, there are no restrictions on who gets to work on tickets. I don't know of anyone else working on this, so feel free to get started. Please keep in mind that changes to existing behavior are disallowed, unless there is a compelling reason for the change.

**Comment by johnjaylward** [comment]
This is not a bug and was the intended way for enums to be serialized and deserialised by this library. This would be a major breaking change for anyone who is using the enum feature

**Comment by ThestralWarrior** [comment]
I understand. It's my first time trying to make an open source contribution. I could use some guidance!

**Comment by stleary** [comment]
@ThestralWarrior No worries, this was an unusual issue since the obvious solution was not right for this project. There will be other opportunities to contribute.

**Comment by robross0606** [comment]
> This is not a bug and was the intended way for enums to be serialized and deserialised by this library. This would be a major breaking change for anyone who is using the enum feature

What is the workaround for this problem?  JSONObject is treating Enum differently than any other class.  If a user wants to serialize and deserialize an enum value using something other than `name`, there should be a way to do that.  If it isn't by overriding `toString()` and `valueOf()` then what is it?

**Comment by srl295** [comment]
@stleary hi, i ran into this when upgrading from a 2002 (yes) version of the library.

You wrote:

> This is how the library was designed

Just as a historical note, the code involved was introduced in 2016 in https://github.com/stleary/JSON-java/pull/271 - originally `toString()` was used (by default), and this was changed to `name()` for Enums.

Originally, .toString() was called on the enum element, which was overridable.

**Comment by johnjaylward** [comment]
Yes, the original method of serialization for enums was broken as noted in the PR. 
As such we designed the bug fix around using the name of the enum.

This was mainly to ensure round tripping of bean -> json -> bean. Using the `toString` method would have made a round trip near impossible to do easily

**Comment by stleary** [comment]
The default behavior won't be changed. No objections if someone wants to add a `JSONParserConfiguration` option to get a different outcome.

**Comment by XIAYM-gh** [comment]
Hello, I'd like to work on this.

But as mentioned by johnjaylward, using `toString()` would make the deserialization process nearly impossible. As far as I'm concerned, it's worthwhile to discuss:

1. Should we keep the `fromJSON()` compatibility?
2. If so, how to implement?

Personally, for the latter question, I'd like to introduce an annotation which decorates a static deserialize method returning the exact enum type it is. For instance:

```java
public enum MyEnum {
    V_1,
    V_2;

    @Override
    public String toString() {
        switch (this) {
            case V_1:
                return "1.0";
        ...
    }

    @JSONEnumDeserializer
    public static MyEnum deserialize(String input) {
        switch (input) {
            case "1.0":
                return MyEnum.V_1;
        ...
    }
}
```

**Comment by johnjaylward** [comment]
I think a better approach would be to add support for generic converters like other languages and libraries have.
`@JsonConverter(MyConverterClass<From,To>.class)`
This would allow us to handle more than just enums in a consistent way. It also makes the JSON processing loser coupled instead of embedding JSON serialization directly into the type definition.

**Comment by XIAYM-gh** [comment]
Wow, that's definitely an extraordinary idea. However, it seems to be a large change, and I think we need more discussion on it, as well as the approval from stleary.

**Comment by stleary** [comment]
@XIAYM-gh Please follow @johnjaylward's guidance

**Comment by XIAYM-gh** [comment]
Copy that. I'll start working on it soon.

**Comment by XIAYM-gh** [comment]
I've met some troubles:

1. On the one hand, creating converter instance every time is inefficient.
2. On the other hand, persisting them in memory is a bad solution.

Personally, I'd prefer using `JSONParserConfiguration`:

```java
JSONParserConfiguration config = new JSONParserConfiguration()
                .withConverter(DataA.class, new MyConverterA())
                .withConverter(DataB.class, new MyConverterB());
```

And the data class:

```java
public class DataA {
    public String data;
}
```

As well as the implementation side:

```java
public class MyConverterA implements JSONConverter<DataA, JSONObject> {
    @Override
    public DataA fromJSON(JSONObject object) { ... }

    @Override
    public JSONObject toJSON(DataA data) { ... }
}
```

What's your opinion on it?
